File: mobx/src/main/java/com/ivansadovyi/mobx/TrackedObservable.java
package com.ivansadovyi.mobx;

import java.util.HashSet;
import java.util.Set;

public class TrackedObservable<T> implements Observable {

	private T value;
	private Set<Observer> observers = new HashSet<>();

	public TrackedObservable(T value) {
		this.value = value;
	}

	public T get() {
		ObservableTracker tracker = ObservableTrackerHolder.get();
		if (tracker != null) {
			tracker.track(this);
		}

		return value;
	}

	public void set(T value) {
		if (this.value != value) {
			this.value = value;
			notifyObservers();
		}
	}

	private void notifyObservers() {
		Set<Observer> observers = new HashSet<>(this.observers);
		for (Observer observer : observers) {
			observer.onChange();
		}
	}

	@Override
	public Disposable subscribe(final Observer observer) {
		observers.add(observer);
		return new Disposable() {
			@Override
			public void dispose() {
				observers.remove(observer);
			}
		};
	}
}

File: mobx/src/main/java/com/ivansadovyi/mobx/TrackedObservable.kt
package com.ivansadovyi.mobx

class TrackedObservable<T>(value: T) : Observable {

	var value: T = value
		get() {
			ActiveObservableTrackerHolder.activeObservableTracker?.track(this)
			return field
		}
		set(value) {
			field = value
			notifyObservers()
		}


	private val observers = mutableSetOf<Observer>()

	override fun observe(observer: Observer): Disposable {
		this.observers.add(observer)
		return Disposable {
			this.observers.remove(observer)
		}
	}

	private fun notifyObservers() {
		this.observers.forEach(Observer::onChange)
	}
}
